File: tests/fixtures/ecosystems/scala_comprehensive/Shapes.scala
package comprehensive

sealed trait Shape {
  def area: Double
  def perimeter: Double
  def describe: String = s"${getClass.getSimpleName}: area=$area"
}

case class Circle(radius: Double) extends Shape {
  override def area: Double = math.Pi * radius * radius
  override def perimeter: Double = 2 * math.Pi * radius
}

case class Rectangle(width: Double, height: Double) extends Shape {
  override def area: Double = width * height
  override def perimeter: Double = 2 * (width + height)
}

case class Triangle(a: Double, b: Double, c: Double) extends Shape {
  override def area: Double = {
    val s = (a + b + c) / 2
    math.sqrt(s * (s - a) * (s - b) * (s - c))
  }
  override def perimeter: Double = a + b + c
}

object Shape {
  def largest(shapes: List[Shape]): Option[Shape] =
    shapes.sortBy(_.area).lastOption
}

File: tests/fixtures/ecosystems/cpp_comprehensive/shapes.cpp
#include "shapes.h"
#include <sstream>

std::string Shape::describe() const {
    return "Shape";
}

Circle::Circle(double radius) : radius_(radius) {}

double Circle::area() const {
    return M_PI * radius_ * radius_;
}

double Circle::perimeter() const {
    return 2 * M_PI * radius_;
}

std::string Circle::describe() const {
    std::ostringstream oss;
    oss << "Circle(r=" << radius_ << ")";
    return oss.str();
}

Rectangle::Rectangle(double width, double height)
    : width_(width), height_(height) {}

double Rectangle::area() const {
    return width_ * height_;
}

double Rectangle::perimeter() const {
    return 2 * (width_ + height_);
}

std::string Rectangle::describe() const {
    std::ostringstream oss;
    oss << "Rectangle(" << width_ << "x" << height_ << ")";
    return oss.str();
}

Square::Square(double side) : Rectangle(side, side) {}

std::string Square::describe() const {
    std::ostringstream oss;
    oss << "Square(" << width_ << ")";
    return oss.str();
}
